src/DotVVM.Framework/Compilation/Static/StaticViewCompiler.cs

```diff
@@ -0,0 +1,277 @@
+using DotVVM.Framework.Compilation.Styles;
+using DotVVM.Framework.Compilation.Validation;
+using DotVVM.Framework.Configuration;
+using DotVVM.Framework.Controls;
+using DotVVM.Framework.Controls.Infrastructure;
+using DotVVM.Framework.Hosting;
+using DotVVM.Framework.Security;
+using DotVVM.Framework.Utils;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CSharp.RuntimeBinder;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+
+namespace DotVVM.Framework.Compilation.Static
+{
+    internal class StaticViewCompiler
+    {
+        public const string ObjectsClassName = "SerializedObjects";
+        private const string IDotvvmCacheAdapterName
+            = "DotVVM.Framework.Runtime.Caching.IDotvvmCacheAdapter, DotVVM.Framework";
+        private const string SimpleDictionaryCacheAdapterName
+            = "DotVVM.Framework.Testing.SimpleDictionaryCacheAdapter, DotVVM.Framework";
+        private readonly ImmutableArray<MetadataReference> baseReferences;
+
+        private readonly DotvvmConfiguration configuration;
+        private readonly Assembly dotvvmProjectAssembly;
+
+        // NB: Currently, an Assembly must be built for each view/markup control (i.e. IControlBuilder) and then merged
+        //     into one assembly. It's horrible, I know, but the compiler is riddled with references to System.Type
+        //     that make it presently impossible to compile it all in one go.
+        private readonly ConcurrentDictionary<string, StaticView> viewCache
+            = new ConcurrentDictionary<string, StaticView>();
+
+        public StaticViewCompiler(DotvvmConfiguration configuration, Assembly dotvvmProjectAssembly)
+        {
+            this.configuration = configuration;
+            this.dotvvmProjectAssembly = dotvvmProjectAssembly;
+            baseReferences = GetBaseReferences(configuration);
+        }
+
+        public static DotvvmConfiguration CreateConfiguration(
+            Assembly dotvvmProjectAssembly,
+            string dotvvmProjectDir)
+        {
+            InitializeDotvvmControls(dotvvmProjectAssembly);
+            return ConfigurationInitializer.GetConfiguration(dotvvmProjectAssembly, dotvvmProjectDir, services =>
+            {
+                services.AddSingleton<IControlResolver, StaticViewControlResolver>();
+                services.TryAddSingleton<IViewModelProtector, FakeViewModelProtector>();
+                services.AddSingleton(new RefObjectSerializer());
+                // Yes, this is here so that there can be a circular dependency in StaticViewControlResolver.
+                // I'm not happy about it, no, but the alternative is a more-or-less complete rewrite.
+                services.AddSingleton(p => new StaticViewCompiler(
+                    p.GetRequiredService<DotvvmConfiguration>(),
+                    dotvvmProjectAssembly));
```

Comment:
How is DotvvmConfiguration dependent on `StaticViewCompiler`?

Reply:
There was the same circular dependency as in `DefaultControlBuilderFactory`. Since `StaticViewCompiler` no longer duplicates `DefaultViewCompiler` this is resolved.